### PR TITLE
Preserve NJT nested ints in AOT subclass outputs

### LINKED ISSUE #137275: Device transfer for NJT within torch.compile allocates a new nested int
Note that this differs from eager behavior for `njt.to(device)`, where the nested int from the input shape is purposefully includes in the output shape, as expected.

Repro:

```python
import torch

def f(nt):
    return nt.to(device="cpu")

compiled_f = torch.compile(f)

nt = torch.nested.nested_tensor([
    torch.randn(2, 5),
    torch.randn(3, 5),
    torch.randn(4, 5),
], layout=torch.jagged, device="cuda")

out = f(nt)
out_compile = compiled_f(nt)
# fails; torch.Size([7, j1]) for out and torch.Size([7, j2]) for out_compile
assert out.shape == out_compile.shape
```

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186225

AOTAutograd unwraps traceable tensor subclass outputs into their inner tensors and later reconstructs the subclass at runtime. For jagged NestedTensor device transfers, the graph copies the offsets or lengths tensor. Because subclass output metadata dropped nested-int size and stride entries, runtime reconstruction associated that copied metadata tensor with a newly allocated nested int, so compiled `nt.to(device="cpu")` returned a shape like `torch.Size([3, j2, 5])` instead of preserving eager's input-derived `j1`.

Carry only input-derived nested ints through AOT subclass output metadata and continue suppressing fresh nested ints created inside the graph. During NestedTensor unflattening, register the real runtime offsets or lengths tensor against the carried nested int so the reconstructed NJT preserves the original symbolic ragged dimension. This keeps in-graph NJT construction from leaking untracked proxy symbols while fixing device-transfer reconstruction.

The alternative would be to special-case `NestedTensor.to()` directly, but the metadata loss happens in the generic AOT subclass output path, so threading the nested-int metadata there fixes the root cause for the reconstruction path.

Fixes #137275
Generated by my agent

Test Plan:
- python - <<'PY'
  import torch

  def f(nt):
      return nt.to(device="cpu")

  compiled_f = torch.compile(f)

  nt = torch.nested.nested_tensor([
      torch.randn(2, 5),
      torch.randn(3, 5),
      torch.randn(4, 5),
  ], layout=torch.jagged, device="cuda")

  out = f(nt)
  out_compile = compiled_f(nt)
  assert out.shape == out_compile.shape
  PY
- python test/test_nestedtensor.py TestNestedTensorSubclassCUDA.test_to_device_compile_preserves_nested_int_cuda_float32
- python test/test_nestedtensor.py TestNestedTensorOpInfoCUDA.test_compile_forward_to_cuda_float32
- python test/test_nestedtensor.py TestNestedTensorSubclassCUDA.test_broadcast_shapes_on_in_graph_constructed_njt_cuda_float32
- python test/functorch/test_subclass_codegen.py
- python test/test_nestedtensor.py TestNestedTensorOpInfoCUDA.test_compile_backward_to_cuda_float32
- python test/functorch/test_aotdispatch.py TestAOTAutograd.test_compute_outer_size_and_stride_bug (blocked locally by torchvision::nms registration failure before test collection)
- runpy/import-shim equivalent of TestAOTAutograd.test_compute_outer_size_and_stride_bug with torchvision disabled
- runpy/import-shim equivalent of TestPartitioning.test_backward_prologue_correctness_with_subclass and TestPartitioning.test_backward_epilogue_correctness_with_subclass with torchvision disabled
- lintrunner -a

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the Dr. CI failures in the NestedTensor CPU jobs. The failures came from fresh in-graph nested ints being admitted into AOT subclass output metadata during joint tracing, plus legitimate input-derived SingletonInt expressions reaching Inductor size hinting; the update narrows the allowlist and fixes SingletonInt/optimization hint handling.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the remaining Dr. CI NestedTensor failures. The update preserves raw symbolic outputs that are both used to reconstruct subclass outputs and saved for backward, and trims metadata-only symbolic gradients before wrapping subclass grad inputs.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I fixed the remaining Dr. CI failure in test_njt_subclass_simple. The failure was a stale inline expected graph: the PR now intentionally preserves the input-derived ragged size and stride-0 metadata for NJT subclass outputs, matching the already-updated from_cat expectation.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and fixed two integration regressions revealed by the rebase: backward NJT gradients now retain the primal nested-int identity, and joint-graph partitioning counts the full mutated subclass output arity including nested size and stride metadata. Cache-disabled CPU mutation and CUDA backward regressions pass, along with focused AOT subclass, Dynamo NJT, synthetic-base/deduplication, SingletonInt, and full lintrunner coverage. A fresh formal review found no actionable issues.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
